unison-cli/src/Unison/Codebase/Editor/HandleInput/Update.hs

```diff
@@ -279,3 +603,24 @@ propagatePatchNoSync ::
 propagatePatchNoSync patch scopePath =
   Cli.time "propagatePatchNoSync" do
     Cli.stepAtNoSync' (Path.unabsolute scopePath, Propagate.propagateAndApply patch)
+
+------------------------------------------------------------------------------------------------------------------------
+-- Tiny helper bijection type
+--
+-- This is semantically a set of `(a, b)` tuples, where no `a` nor `b` appears more than once. An `a` can be looked up
+-- given its associated `b`, and vice-versa.
+
+data Bij a b
```

Comment:
Don't have to do this now, but this seems like a useful type; I'd be in favour of moving it to a util package at some point, or just lumping it into `util-relation` as a new module.

Reply:
I'm not in favor of moving it out yet; it's really just to locally reduce the number of random mappy/setty data structures in this algorithm by one.